### PR TITLE
#1199 Followup

### LINKED ISSUE #1308: Fix (+test) phantom node onion error returning
See https://github.com/lightningdevkit/rust-lightning/pull/1199#discussion_r806303997 and the following comment, probably need to test al the cases where we return onion errors in that function. Also some nits worth hitting at https://github.com/lightningdevkit/rust-lightning/pull/1199#pullrequestreview-882275581

### PR DESCRIPTION
**Pull request by valentinewallace**
Closes #1308.

### CODE REVIEW

File: lightning/src/ln/channelmanager.rs [hunk 1]
```diff
@@ -3011,45 +3011,62 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
...
 														let mut arr = [0; 32];
 														arr.copy_from_slice(&SharedSecret::new(&onion_packet.public_key.unwrap(), &phantom_secret_res.unwrap())[..]);
 														arr
 													};
-													let next_hop = match onion_utils::decode_next_hop(shared_secret, &onion_packet.hop_data, onion_packet.hmac, payment_hash) {
+													let next_hop = match onion_utils::decode_next_hop(phantom_shared_secret, &onion_packet.hop_data, onion_packet.hmac, payment_hash) {
 														Ok(res) => res,
 														Err(onion_utils::OnionDecodeErr::Malformed { err_msg, err_code }) => {
-															fail_forward!(err_msg, err_code, Vec::new());
+															fail_phantom_forward!(err_msg, err_code, Vec::new(), phantom_shared_secret);
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmanager.rs:3059]: Hmm, shouldn't malformed not get encrypted by the phantom node? Malformed generally is "send it back to previous node and let them encrypt it".

Comment (valentinewallace) [reply lightning/src/ln/channelmanager.rs:3059]: Oops, will fix

File: lightning/src/ln/channelmanager.rs [hunk 2]
```diff
@@ -3011,45 +3011,63 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 									HTLCForwardInfo::AddHTLC { prev_short_channel_id, prev_htlc_id, forward_info: PendingHTLCInfo {
 										routing, incoming_shared_secret, payment_hash, amt_to_forward, outgoing_cltv_value },
 										prev_funding_outpoint } => {
+											macro_rules! get_failure_htlc_source {
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmanager.rs:3014]: Cant this just be a variable instead of a macro?

**Comment by TheBlueMatt** [comment]
Can you add some more tests - eg (a) that don't modify the onion at all, by just editing the route to have incorrect values - too-low CLTV, too low amount, etc (b) that hit `blamed_chan_closed(false)`, (c) that also blame the prev SCID not just the phantom SCID?

**Comment by valentinewallace** [comment]
> Can you add some more tests - eg (a) that don't modify the onion at all, by just editing the route to have incorrect values - too-low CLTV, too low amount, etc (b) that hit `blamed_chan_closed(false)`, (c) that also blame the prev SCID not just the phantom SCID?

Thanks, I think I added these but whew each one found a bug basically (one preexisting tbf). So let me know if anything's missing

**Comment by TheBlueMatt** [comment]
I'm not quite sure why the test I wrote at https://git.bitcoin.ninja/index.cgi?p=rust-lightning;a=commit;h=1907bcc5d055008eae4e378f2907cba15d719587 is failing - can you examine it a bit and include that test here?

**Review by jkczyz — APPROVED** [review]

**Pull request merged**